sensors/cmd/start.go

```diff
@@ -70,20 +45,194 @@ func Start() {
+	}
+
+	for {
+		select {
+		case f := <-manager.C:
+			if err := f(); err != nil {
+				logger.Fatalw("failed to listen to events", zap.Error(err))
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+type SensorContextManager struct {
+	C               chan func() error
+	logger          *zap.SugaredLogger
+	kubeClient      kubernetes.Interface
+	dynamicClient   dynamic.Interface
+	eventBusConfig  *eventbusv1alpha1.BusConfig
+	eventBusSubject string
+	hostname        string
+}
+
+func NewSensorContextManager(logger *zap.SugaredLogger, kubeClient kubernetes.Interface, dynamicClient dynamic.Interface, eventBusConfig *eventbusv1alpha1.BusConfig, eventBusSubject, hostname string) *SensorContextManager {
+	return &SensorContextManager{
+		C:               make(chan func() error),
+		logger:          logger,
+		kubeClient:      kubeClient,
+		dynamicClient:   dynamicClient,
+		eventBusConfig:  eventBusConfig,
+		eventBusSubject: eventBusSubject,
+		hostname:        hostname,
+	}
+}
+
+func (scm *SensorContextManager) Start(ctx context.Context) error {
+	// SENSOR_PATH env variable is used to indicate if this sensor definition
+	// has opted in to live reload
+	path, liveReload := os.LookupEnv("SENSOR_PATH")
+
+	switch liveReload {
+	case true:
+		return scm.startLiveReload(ctx, path)
+	case false:
+		return scm.startDefault(ctx)
+	}
+
+	return nil
+}
+
+func (scm *SensorContextManager) Close() {
+	close(scm.C)
+}
+
+func (scm *SensorContextManager) startDefault(ctx context.Context) error {
+	sensor, err := scm.readSensorFromEnv()
+	if err != nil {
+		scm.logger.Fatalw("failed to read sensor from env variable", zap.Error(err))
 	}
 
-	ctx := logging.WithLogger(signals.SetupSignalHandler(), logger)
 	m := metrics.NewMetrics(sensor.Namespace)
 	go m.Run(ctx, fmt.Sprintf(":%d", common.SensorMetricsPort))
 
-	logger.Infow("starting sensor server", "version", argoevents.GetVersion())
-	sensorExecutionCtx := sensors.NewSensorContext(kubeClient, dynamicClient, sensor, busConfig, ebSubject, hostname, m)
-	if err := sensorExecutionCtx.Start(ctx); err != nil {
-		logger.Fatalw("failed to listen to events", zap.Error(err))
+	sensorCtx := sensors.NewSensorContext(scm.kubeClient, scm.dynamicClient, sensor, scm.eventBusConfig, scm.eventBusSubject, scm.hostname, m)
+	scm.C <- func() error { return sensorCtx.Start(ctx) }
+
+	return nil
+}
+
+func (scm *SensorContextManager) startLiveReload(ctx context.Context, path string) error {
+	// watch for sensor file changes
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return err
+	}
+
+	err = watcher.Add(path)
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		defer watcher.Close()
+		var m *metrics.Metrics
+
+		for {
+			sensor, err := scm.readSensorFromFile(path)
+			if err != nil {
+				scm.logger.Fatalw("failed to read sensor from file", zap.Error(err))
+			}
+
+			// logger
+			scm.logger = scm.logger.With("sensorName", sensor.Name)
```

Comment:
Lines 156-159 to add key value pairs to the logger need to be added to the default sensor.

Reply:
Nice catch, [added](https://github.com/intuit-data-os/argo-events/pull/29/files#diff-7c6e3f9dab2b101aae1efe37888503488a33f5705363b48d47a6533f56e1ec85R211-R215)!